### DISCUSSION TITLE
Optional new() constraint

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2483] [Q&A] [answered]
Hi All,
I'm looking for some advice on how to implement something in C#.
TLDR:  I want to write code like this:
````
public abstract class Base<TProvider> {

  protected virtual TProvider CreateProvider() when TProvider:new() {
    return new TProvider();
  }

  protected abstract TProvider CreateProvider() when ! TProvider:new();

}
````

Today, I implement it as follows:
````
public abstract class Base0<TProvider> {
  protected abstract TProvider CreateProvider();
}

public abstract class Base1<TProvider> : Base0<TProvider> where TProvider : new(){
  protected override TProvider CreateProvider() {
    return new TProvider();
  }
}
````

But the problem is that as create more enhanced bases, have to have two versions of each.  One for T:new() and one for !T:new().

What is a good way for me to have a single generic class in the spirit of my ideal at the top?

**Comment by HaloFour** [comment]
The CLR doesn't recognize generic type constraints as a part of the method signature so you are not able to overload in such a manner.  I'd suggest using two different method names.

**Comment by john-h-k** [comment]
Basically, this is overloading by generic constraint. Not supported currently by the CLR currently, but could be very useful feature

**Comment by jnm2** [comment, marked as answer]
`new TProvider()` is syntax sugar for `Activator.CreateInstance<TProvider>()`. You can do the same with reflection to test whether a public parameterless constructor exists.
